lib/Dialects/LinalgExt/Transforms/TileToInParallel.cpp

```diff
@@ -0,0 +1,178 @@
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/AffineExpr.h"
+#include "mlir/IR/Identifier.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace mlir;
+using namespace mlir::linalg_ext;
+
+namespace {
+
+struct Tie {
+  AffineExpr e;
+  Value v;
+  operator AffineExpr() const { return e; }
+  operator Value() const { return v; }
+};
+
+/// Helper struct to build simple arithmetic quantities with minimal type
+/// inference support.
+// TODO: move into ArithBuilder once ops have been moved into arith.
+struct AffineBuilder {
+  AffineBuilder(OpBuilder &b, Location loc) : b(b), loc(loc) {}
+
+  Value add(Tie lhs, Tie rhs) {
+    return b.createOrFold<AffineApplyOp>(
+        loc, ArrayRef<AffineExpr>{lhs.e + rhs.e}, ValueRange{lhs, rhs});
+  }
+  Value sub(Tie lhs, Tie rhs) {
+    return b.createOrFold<AffineApplyOp>(
+        loc, ArrayRef<AffineExpr>{lhs.e - rhs.e}, ValueRange{lhs, rhs});
+  }
+  Value mul(Tie lhs, Tie rhs) {
+    return b.createOrFold<AffineApplyOp>(
+        loc, ArrayRef<AffineExpr>{lhs.e * rhs.e}, ValueRange{lhs, rhs});
+  }
+  Value ceil(Tie lhs, Tie rhs) {
+    return b.createOrFold<AffineApplyOp>(
+        loc, ArrayRef<AffineExpr>{lhs.e.ceilDiv(rhs.e)}, ValueRange{lhs, rhs});
+  }
+  Value min(ValueRange vals) {
+    return b.createOrFold<AffineMinOp>(
+        loc, AffineMap::getMultiDimIdentityMap(vals.size(), b.getContext()),
+        vals);
+  }
+  Value max(ValueRange vals) {
+    return b.createOrFold<AffineMinOp>(
+        loc, AffineMap::getMultiDimIdentityMap(vals.size(), b.getContext()),
+        vals);
+  }
+
+ private:
+  OpBuilder &b;
+  Location loc;
+};
+
+struct TileOpToInParallelRewriter
+    : public OpRewritePattern<linalg_ext::TileOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg_ext::TileOp tileOp,
+                                PatternRewriter &rewriter) const override {
+    // TODO: verifier.
+    assert(tileOp.getNumResults() > 0 &&
+           tileOp.outs().size() == tileOp.getNumResults());
+
+    // TODO: when supported, iterate over the tensor of sizes. This will be
+    // iterating through a level of indirection.
+
+    // Construct the loop bounds based on the canonical arithmetic progression.
+    Location loc = tileOp.getLoc();
+    Value zero = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+    Value one = rewriter.create<arith::ConstantIndexOp>(loc, 1);
+    Value totalSize =
+        rewriter.create<tensor::DimOp>(loc, tileOp.outs().front(), zero);
+    Value step = tileOp.tile_sizes();
+    assert(step.getType().isa<IndexType>() && "NYI: not an index type");
+
+    AffineBuilder ab(rewriter, loc);
+    AffineExpr i, j, M;
+    bindDims(rewriter.getContext(), i, j);
+    bindSymbols(rewriter.getContext(), M);
+    Value sizePerThread = ab.ceil(Tie{i, totalSize}, Tie{M, step});
```

Comment:
It's not really `sizePerThread`, right? It's more like `numThreads` or `numTiles`. The previous name was a little confusing for me